experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,697 @@
+
+  ControlState(TypedValue<handshake::ControlType> control)
+      : TypedValueState<handshake::ControlType>(control) {}
+
+protected:
+  bool valid = false;
+  bool ready = false;
+};
+
+//===----------------------------------------------------------------------===//
+// Updater
+//===----------------------------------------------------------------------===//
+/// Classes to update old values states with new ones after finishing the data
+/// collection.
+/// The map updaters (defined inside the Simulator class) contains references to
+/// the corresponding pair of new and old valueStates, assigns the newly
+/// collected values from newValueStates to oldValueStates when the collection
+/// is finished.
+
+/// Base Updater class
+class Updater {
+public:
+  virtual ~Updater() = default;
+  virtual void update() = 0;
+};
+
+/// Templated updater class that contains pair <oldValueState, newValueState>
+template <typename State>
+class DoubleUpdater : public Updater {
+protected:
+  State &cons;
+  State &prod;
+
+public:
+  DoubleUpdater(State &cons, State &prod) : cons(cons), prod(prod) {}
+};
+
+/// Update valueStates with "Channel type": valid, ready, data
+class ChannelUpdater : public DoubleUpdater<ChannelState> {
+public:
+  ChannelUpdater(ChannelState &cons, ChannelState &prod)
+      : DoubleUpdater<ChannelState>(cons, prod) {}
+  void update() override {
+    cons.valid = prod.valid;
+    cons.ready = prod.ready;
+    cons.data = prod.data;
+  }
+};
+
+/// Update valueStates with "Control type": valid, ready
+class ControlUpdater : public DoubleUpdater<ControlState> {
+public:
+  ControlUpdater(ControlState &cons, ControlState &prod)
+      : DoubleUpdater<ControlState>(cons, prod) {}
+  void update() override {
+    cons.valid = prod.valid;
+    cons.ready = prod.ready;
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Readers & writers
+//===----------------------------------------------------------------------===//
+
+/// Classes with references to oldValueStates and newValueStates to represent
+/// readers/writers API (that is opened to access for the user). The map<
+/// <value, Operation*>, RW*> rws (inside the simulator class) stores these
+/// classes and is deleted by the simulator itself. In fact, RW is a sort of the
+/// union of the StateReader and StateWriter classes from your code
+
+/// Base RW
+/// Virtual destructor is not necessary but maybe add some virtual func members
+/// later
+class RW {
+public:
+  RW(Value val) : val(val) {}
+  virtual ~RW() {};
+
+private:
+  Value val;
+};
+
+/// Templated RW
+template <typename Ty>
+class TypedRW : public RW {
+public:
+  TypedRW(TypedValue<Ty> val) : RW(val) {}
+};
+
+////--- Channel (data)
+
+/// In this case the user can change the valid signal, but has ReadOnly access
+/// to the valid and data ones.
+struct ChannelConsumerRW : public TypedRW<handshake::ChannelType> {
```

Comment:
I like these dual reader/writer for bi-directional values, they are less annoying than my 4 separate readers/writers!